[FEATURE][processing] Snap geometries to layer algorithm

nyalldawson opened the pull request:
This PR ports the Geometry Snapper plugin across to the analysis lib, and exposes it to python bindings

A new processing algorithm has been added which performs the snapping to layers.

Next step would be removing the current c++ plugin, since 99% of it's functionality is now present in processing. The only missing bits are that the plugin could run in the background - but given that task manager is on it's way I don't think this is a big loss, and the gains are significant (using snapping in processing models, one less c++ plugin).

@manisandro What's your thoughts?

nyalldawson committed:
[FEATURE][processing] Snap geometries to layer algorithm

Port the Geometry Snapper plugin across to the analysis lib, and
expose to python bindings

Add a new algorithm which performs the snapping to layers

m-kuhn commented on python/analysis/vector/qgsgeometrysnapper.sip:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * Snaps a set of features to the reference layer and returns the result. This operation is
+     * multithreaded for performance. The featureSnapped() signal will be emitted each time a feature
+     * is processed. This method is not safe to call from multiple threads concurrently.
```
I wonder if we should stick to the Qt terminology for thread-safety:

http://doc.qt.io/qt-4.8/threads-reentrancy.html

> Throughout the documentation, the terms reentrant and thread-safe are used to mark classes and functions to indicate how they can be used in multithread applications:
* A thread-safe function can be called simultaneously from multiple threads, even when the invocations use shared data, because all references to the shared data are serialized.
* A reentrant function can also be called simultaneously from multiple threads, but only if each invocation uses its own data.

nyalldawson replied:
Ah agreed - however that comment was outdated and has been removed anyway :)

manisandro commented:
@nyalldawson LGTM

nyalldawson committed:
[FEATURE][processing] Snap geometries algorithm allows snapping
to other layer types, supports point/line layers

Fix #14791, #15313

nyalldawson committed:
Port some API from QgsPoint to QgsPointV2

nyalldawson committed:
[FEATURE] Snap to layer algorithm accepts a mode parameter

With a new option to prefer to snap to closest point on geometry.
The old behaviour was to prefer to snap to nodes, even if a node
was further from the input geometry than a segment. The new option
allows you to snap geometries to the closest point, regardless
of whether it's a node or segment.

nyalldawson committed:
[FEATURE] Remove c++ geometry snapper plugin

All functionality is now available through analysis lib + processing
algorithm.

Marked as feature for documentation + changelog flagging
